Fix launch grid calculation

nandesuka opened the pull request:
Summary: 

The launch grid calculation code is using a python trick to achieve CeilDiv() through negative integer division with FloorDiv(). This is language dependent behaviour that doesn't apply to all languages.

In the FXIR backend we negate this behaviour and replace the experssion with CeilDiv() operation so the computation is correct regardless of language used. Not directly directly changing the orginal computation as it leads to a performance degredation.

Test Plan:
CI

Rollback Plan:

Differential Revision: D79275534


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

blaine-rister reviewed (commented):
Mostly LGTM, left a few comments.

blaine-rister commented on torch/_inductor/codegen/wrapper_fxir.py:
```diff
@@ -581,8 +581,8 @@ def replace_floor_div(expr: sympy.Expr) -> sympy.Expr:
-
-                return FloorDiv(numerator, denominator)
+                # Undo the python division trick and replace with explict CeilDiv
+                return -CeilDiv(-numerator, denominator)
```
Just out of curiosity, how does this change the final expressions appearing in the GraphModule? One of the nice things about `FloorDiv` is it maps to Python's built-in `//`, but it's possible that `CeilDiv` works as well. Could you show the output of `print(gm)` on a GM with dynamic shapes before and after this change?

nandesuka committed:
Fix launch grid calculation (#159497)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159497

The launch grid calculation code is using a python trick to achieve CeilDiv() through negative integer division with FloorDiv(). This is language dependent behaviour that doesn't apply to all languages.

In the FXIR backend we negate this behaviour and replace the experssion with CeilDiv() operation so the computation is correct regardless of language used. Not directly directly changing the orginal computation as it leads to a performance degredation.

Test Plan:
CI

Rollback Plan:

Reviewed By: jfix71

Differential Revision: D79275534

blaine-rister reviewed (approved):
LGTM.

blaine-rister commented:
@pytorchbot merge
